format-avro/src/main/java/io/cdap/plugin/format/avro/AvroToStructuredTransformer.java

```diff
@@ -91,9 +95,76 @@ public Schema convertSchema(org.apache.avro.Schema schema) throws IOException {
     if (schemaCache.containsKey(hashCode)) {
       structuredSchema = schemaCache.get(hashCode);
     } else {
-      structuredSchema = Schema.parseJson(schema.toString());
+      String strSchema = schema.toString();
+      String jsonSchema = preprocessSchema(new Gson().fromJson(strSchema, JsonObject.class)).toString();
+      structuredSchema = Schema.parseJson(jsonSchema);
       schemaCache.put(hashCode, structuredSchema);
     }
     return structuredSchema;
   }
+
+  private JsonObject preprocessSchema(JsonObject schema) {
+
+    if (!schema.has("type")) {
+      return schema;
+    }
+
+    JsonElement type = schema.get("type");
+
+    if (type.isJsonArray()) {   // Union
+      for (JsonElement subtype : type.getAsJsonArray()) {
+        if (!subtype.isJsonPrimitive()) {
+          preprocessSchema(subtype.getAsJsonObject());
+        } else if (customTypes.containsKey(subtype.getAsString())) {
+          schema.remove("type");
+          schema.add("type", customTypes.get(subtype.getAsString()));
+        }
+      }
+    } else if (type.isJsonObject()) {   // Unnamed Complex type
+      preprocessSchema(type.getAsJsonObject());
+    } else {
+      String typeName = type.getAsString();
+      switch (typeName) {
+        case "record":
+          if (!schema.has("fields")) {
+            return schema;
+          }
+          for (JsonElement field : schema.get("fields").getAsJsonArray()) {
+            preprocessSchema(field.getAsJsonObject());
+          }
+          customTypes.put(schema.getAsJsonPrimitive("name").getAsString(), schema);
+          break;
+
+        case "map":
+          schema.addProperty("keys", "string");
+          if (!schema.get("values").isJsonPrimitive()) {
+            preprocessSchema(schema.getAsJsonObject("values"));
+          }
+          break;
+
+        case "array":
+          JsonElement items = schema.get("items");
+          if (!items.isJsonPrimitive()) {
+            preprocessSchema(schema.getAsJsonObject("items"));
+          } else if (customTypes.containsKey(items.getAsString())) {
+            schema.remove("items");
+            schema.add("items", customTypes.get(items.getAsString()));
```

Comment:
So, you are replacing type references with named type definitions. Why? Could you please explain in the PR description what is it doing and why. As of now, it would potentially create tons of duplicates of named type definitions within schema that share same name.

Reply:
@fernst could you please review this as well? I am not sure about duplicating records with same name

Reply:
This is to handle the point "No support for referencing custom types" mentioned in the PR description above. If this is not done, the CDAP schema attempts to parse the type reference using one of the built in types, which it does not find, leading to an error.

For example, if an enum named `myenum` is already defined and referenced in another field as `{"name": "foo", "type": "myenum"}`, it throws an error `No enum constant io.cdap.cdap.api.data.schema.Schema.Type.MYENUM` as it is not present in [Schema.Type](https://github.com/cdapio/cdap/blob/develop/cdap-api-common/src/main/java/io/cdap/cdap/api/data/schema/Schema.java#L57-L72).

Reply:
Should you then remove "name" when doing substitutions? Or set unique name for each substitution? Otherwise we have many copies with same name. I believe @fernst was solving similar problem